lib/defaultingress/api_ingress.go

```diff
@@ -0,0 +1,292 @@
+	result, err = controllerutil.CreateOrUpdate(context.TODO(), c, ingress, func() error {
+		if e := controllerutil.SetControllerReference(cr, ingress, scheme); e != nil {
+			return e
+		}
+
+		ingress.SetAnnotations(annotations)
+		ingress.Spec = ingressSpec
+
+		return nil
+	})
+	if err == nil {
+		misc.LogCreateOrUpdateOperationResult(log, result, cr, ingress)
+	}
+
+	return err
+}
+
+func getAPIIngressName(cr *ingressv1alpha1.AstarteDefaultIngress) string {
+	return cr.Name + "-api-ingress"
+}
+
+func getConfigMapName(cr *ingressv1alpha1.AstarteDefaultIngress) string {
+	return cr.Name + "-api-ingress-config"
+}
+
+func getAPIIngressAnnotations(cr *ingressv1alpha1.AstarteDefaultIngress) map[string]string {
+	annotations := map[string]string{
+		"nginx.ingress.kubernetes.io/ssl-redirect":   "false",
+		"nginx.ingress.kubernetes.io/use-regex":      "true",
+		"nginx.ingress.kubernetes.io/rewrite-target": "/$2",
+		"nginx.ingress.kubernetes.io/configuration-snippet": "more_set_headers \"X-Frame-Options: SAMEORIGIN\";\n" +
+			"more_set_headers \"X-XSS-Protection: 1; mode=block\";\n" +
+			"more_set_headers \"X-Content-Type-Options: nosniff\";\n" +
+			"more_set_headers \"Referrer-Policy: no-referrer-when-downgrade\";",
+	}
+
+	// Should we serve /metrics?
+	if !pointy.BoolValue(cr.Spec.API.ServeMetrics, false) {
+		allowSubnetAnnotation := ""
+		if cr.Spec.API.ServeMetricsToSubnet != "" {
+			allowSubnetAnnotation = fmt.Sprintf("allow %s;\n", cr.Spec.API.ServeMetricsToSubnet)
+		}
+
+		serverSnippetValue := fmt.Sprintf("location ~* \"/(appengine|flow|housekeeping|pairing|realmmanagement)/metrics\" {\n"+
+			"%s"+
+			"  deny all;\n"+
+			"  return 404;\n"+
+			"}", allowSubnetAnnotation)
+
+		annotations["nginx.ingress.kubernetes.io/server-snippet"] = serverSnippetValue
+	}
+
+	// Should we enable cors?
+	if pointy.BoolValue(cr.Spec.API.Cors, false) {
+		annotations["nginx.ingress.kubernetes.io/enable-cors"] = strconv.FormatBool(true)
+	}
+
+	return annotations
+}
+
+func getAPIIngressSpec(cr *ingressv1alpha1.AstarteDefaultIngress, parent *apiv1alpha1.Astarte) networkingv1.IngressSpec {
+	ingressSpec := networkingv1.IngressSpec{
+		// define which ingress controller will implement the ingress
+		IngressClassName: getIngressClassName(cr),
+		TLS:              getAPIIngressTLS(cr, parent),
+		Rules:            getAPIIngressRules(cr, parent),
+	}
+
+	return ingressSpec
+}
+
+func getIngressClassName(cr *ingressv1alpha1.AstarteDefaultIngress) *string {
+	if cr.Spec.IngressClass == "" {
+		return pointy.String("nginx")
+	}
+	return pointy.String(cr.Spec.IngressClass)
+}
+
+func getAPIIngressTLS(cr *ingressv1alpha1.AstarteDefaultIngress, parent *apiv1alpha1.Astarte) []networkingv1.IngressTLS {
+	ingressTLSs := []networkingv1.IngressTLS{}
+
+	// Check API
+	if pointy.BoolValue(parent.Spec.API.SSL, true) || pointy.BoolValue(cr.Spec.Dashboard.SSL, true) {
+		var secretName string
+
+		// Check the API Spec
```

Comment:
Isn't this just a bit too verbose (including the variable declaration)? Think about this:

```
secretName := cr.Spec.TLSSecret
if cr.Spec.API.TLSSecret != "" {
	secretName = cr.Spec.API.TLSSecret
}
```